### ISSUE TITLE
Akka HTTP and HttpURLConnection - Responding to requests early

### ISSUE THREAD

**Issue by RichardBradley** [akka/akka#16893]
There is a nasty interaction between Java's HttpURLConnection and Akka's HTTP server if you respond to a request without having read the full HTTP request body.

I think that all HTTP servers ought to always either read in the full request and/or half-close the TCP upload stream.

I think that Akka HTTP ought to help apps by making it easy to do that.

This thread on akka-user has lots more details: https://groups.google.com/forum/#!topic/akka-user/MK8Ug9vV2K4

This GitHub repos has a failing test which demonstrates the issue: 
https://github.com/RichardBradley/akka-connection-reuse/blob/3ce125558/src/test/scala/myproject/EarlyHttpResponsesSpec.scala

(I tried to make this a "gist" rather than a repos, but GitHub rejected my push with " ! [remote rejected] master -> master (pre-receive hook declined)
" -- does anyone know why? Maybe there's a max file count on a gist? I can't find it documented...)

**Comment by RichardBradley** [comment]
Things to note in that repro case:

* The [EarlyHttpResponsesSpec](https://github.com/RichardBradley/akka-connection-reuse/blob/master/src/test/scala/myproject/EarlyHttpResponsesSpec.scala) demonstrates the issue
 * The requests will time out once the server gets into this invalid state.
 * There are some shenanigans needed in that test in order to make multiple requests on the same connection.
 * The requests ought to all complete, I think.
* In the linked akka-user thread, it was suggested that the server should "cancel" the stream using `request.entity.dataBytes.runWith(Sink.cancelled)` rather than consume the stream. 
 * I found that didn't fix the issue, I needed to read in the full request, not cancel the stream.
 * You can try that by changing "Sink.ignore" to "Sink.cancelled" on line [124 of HttpServer](https://github.com/RichardBradley/akka-connection-reuse/blob/master/src/main/scala/myproject/HttpServer.scala#L124) -- it makes the problem worse, not better :-(
 * I think this represents a bug in Java's HttpURLConnection -- it is re-using a TCP stream when that stream is still in use by a previous connection
  * Is there anything I can do, as an Akka HTTP server, to mitigate that bug?
* The server can recover from part of this bug by consuming the request body, using "runWith(Sink.ignore)"
 * I added a demo of this to the HttpServer class.
 * However, I can't see any way that an Akka HTTP server can cancel or consume request streams which have been part-consumed by endpoint code.
 * See [line 100 oif HttpServer](https://github.com/RichardBradley/akka-connection-reuse/blob/master/src/main/scala/myproject/HttpServer.scala#L100)

You may argue that the real bug here is endpoint code part-consuming a request stream, and not anything that Akka is doing wrong.
The "endpoint" code is a bit contrived in this test case, I agree.

We are running into this in practice with endpoints that deal with multipart MIME uploads. Using Akka's multipart unmarshallers, we use "prefixAndTail" to read the first MIME part (our upload metadata). We then want to either consume the rest of the upload, or reject the upload at that point in the stream.

**Comment by RichardBradley** [comment]
In the short-term, what should my app code do when it wants to cancel a multipart MIME upload after the first body part has been read?

This test seems to show that I can't "cancel" the stream when the client is an instance of Java's HttpURLConnection :-(

**Comment by RichardBradley** [comment]
... at the risk of confusing matters by restating the above discussion, I have attempted to summarise what I think are the bugs here:

There are 3 main cases re replying to HttpRequests:

1. Read only the request headers (not the body), then reply early
   * The server ought to either consume the upload or close the upload stream (close is much preferable, for efficiency)
      * BUG: Akka does not do either of these by default (see [EarlyHttpResponsesSpec.scala:78](https://github.com/RichardBradley/akka-connection-reuse/blob/master/src/test/scala/myproject/EarlyHttpResponsesSpec.scala#L78) in the attached example)
      * Note that this doesn't even involve the "routing" / "endpoint" distinction I have tried to make here -- if the routing DSL logic tests only the headers then rejects the request, Akka still doesn't consume or close the request body.
   * The client ought to notice that the upload stream is closed
      * BUG: java.net.HttpURLConnection does not notice this, and puts the stream in the pool for reuse
      * Can Akka work around this restriction? It can by consuming the request rather than closing it. I have done that in the attached example - see [HttpServer.scala:120](https://github.com/RichardBradley/akka-connection-reuse/blob/master/src/main/scala/myproject/HttpServer.scala#L120)
2. Read part of the body, then reply early
   * The server ought to either consume the upload or close the upload stream
      * BUG (arguable?): Akka does not either of these by default (see [EarlyHttpResponsesSpec.scala:64](https://github.com/RichardBradley/akka-connection-reuse/blob/master/src/test/scala/myproject/EarlyHttpResponsesSpec.scala#L64) in the attached example)
      * You may argue that it is the responsibility of the "endpoint" code to not leave the body part-consumed
      * I think that user code ought to be able to abort (say throw an exception) and the framework should recover here, i.e. we shouldn't need to litter user code with "try {} finally { source.runWith(Sink.ignore) }" or similar
      * I have added code to the attached example demonstrating that Akka could detect that the body is part-read, but I can't figure out how to cancel or consume it from outside the stream execution context -- see [HttpServer.scala:100](https://github.com/RichardBradley/akka-connection-reuse/blob/master/src/main/scala/myproject/HttpServer.scala#L100)
   * The client ought to notice the upload stream is closed
      * BUG: java.net.HttpURLConnection does not notice this, and puts the stream in the pool for reuse
      * Can Akka work around this?
3. Read the full request, then reply
   * Everything works here, as you might hope :-)
      * See [EarlyHttpResponsesSpec.scala:52](https://github.com/RichardBradley/akka-connection-reuse/blob/master/src/test/scala/myproject/EarlyHttpResponsesSpec.scala#L52) in the attached example

**Comment by ktoso** [comment]
Thanks for reporting!
// cc @sirthias @jrudolph

**Comment by sirthias** [comment]
@RichardBradley Thanks for triggering this discussion (both on the ML as well as here).
As has already been mentioned on the ML the HTTP spec defines a "standard" way to deal with large uploads that might be rejected by the server: The `Expect: 100-continue` header.
If you want to be clean and standard-compliant you could simply require that all large requests to a certain endpoint come with a `Expect: 100-continue` header, e.g. by responding with a 400 response if the request **does not** contain this header.

If you want to solve the problem *without* 100-continue you are in unchartered waters, as you can see from the `java.net.HttpURLConnection` implementation. There appear to be many clients that simply do not start reading from the connection before the request has been fully sent.
There is no way to work with these clients without actually consuming the request fully.

HTTP has pretty clear "send request, then read response" semantics. If you send an "early response", i.e. one that goes out across the wire before the request has been fully sent you don't know whether the client will be able to read it. To me the best way is to simply try it (i.e. send a response) and then close the connection. With akka-http you can do this by sending a response with an explicitly added `Connection: close` header.

There is one thing that this discussion shows: we should implement default detection and handling for early responses. I'll add a ticket for this and report back here.

**Comment by RichardBradley** [comment]
> the HTTP spec defines a "standard" way to deal with large uploads that might be rejected by the server: The Expect: 100-continue header.

"Expect:100" is not suitable when the rejection comes part-way through the request body (e.g. a large multipart MIME upload where we can decide whether to read parts 2 - N after having read part 1, such as [SOAP MTOM uploads](http://en.wikipedia.org/wiki/Message_Transmission_Optimization_Mechanism)).

> HTTP has pretty clear "send request, then read response" semantics. If you send an "early response", i.e. one that goes out across the wire before the request has been fully sent you don't know whether the client will be able to read it.

I don't think this is true. Do you have a spec cite which clearly says "send request, then read response"?

As I noted on the ML, the HTTP specs definitely seem to allow the server to reply early and close the upload stream while the client is doing a large upload, and the client should listen for such a close:
http://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html#sec8.2.2

```"An HTTP/1.1 (or later) client sending a message-body SHOULD monitor the network connection for an error status while it is transmitting the request."```

We can also look to other widely-accepted server implementations like Apache AXIS Tomcat which follows this approach when handling SOAP MTOM uploads.

Would it help if I set up a publically accessible demo of Apache AXIS Tomcat doing an early reply, for experimentation and comparison?

**Comment by sirthias** [comment]
> "Expect:100" is not suitable when the rejection comes part-way through the request body

Yes, that is correct.

>> HTTP has pretty clear "send request, then read response" semantics. If you send an "early
>> response", i.e. one that goes out across the wire before the request has been fully sent you
>> don't know whether the client will be able to read it.
>
> I don't think this is true. Do you have a spec cite which clearly says "send request, then read response"?

The older as well as the current HTTP spec make clear that the client is not REQUIRED to listen while sending. As I said, there are many client implementations that simply do not start reading from the connection before the request has been fully sent.
And as you quote from the outdated RFC 2616 (the new one still contains the same sentence here: http://tools.ietf.org/html/rfc7230#section-6.5) the client SHOULD listen. There is no MUST here.

So, there is no robust solution to this problem, only a best effort approach.
Now, you are right that there are also a number of clients that DO listen while sending. That's why I'd recommend to send an early response with a `Connection: close` header. Reusing the connection after an early response is definitely impossible anyway (and actually forbidden by the spec).

**Comment by RichardBradley** [comment]
>  I'd recommend to send an early response with a Connection: close header. Reusing the connection after an early response is definitely impossible anyway (and actually forbidden by the spec).

That sounds ideal, but how do I detect in my akka.http.server.ExceptionHandler interface (or nearby), whether I need to add that header to a response (i.e. whether the request body has been read)?

I can probably do something along those lines using the "RequestStreamMaterializationMonitorStage" helper in [HttpServer.scala:66](https://github.com/RichardBradley/akka-connection-reuse/blob/master/src/main/scala/myproject/HttpServer.scala#L66). I'll update the test to try that and report back on the results.

Given a) that it's always the right thing to do and b) that it's difficult to achieve in user code, shouldn't Akka do that for me?

I think that the "Connection: close" header will probably fix (work around) the bug in java.net.HttpURLConnection -- I may have been over-thinking this.

**Comment by sirthias** [comment]
> Given a) that it's always the right thing to do and b) that it's difficult to achieve in user code, shouldn't Akka do that for me?

Yes, that's the ticket that I'm currently working on. Will post the link here soon.

**Comment by rkuhn** [comment]
see also https://github.com/akka/akka/issues/16870

**Comment by sirthias** [comment]
This the new ticket https://github.com/akka/akka/issues/16898.
If @RichardBradley agrees then we can close this ticket here.

**Comment by RichardBradley** [comment]
> If @RichardBradley agrees then we can close this ticket here.

I'll retest to confirm that Java's HttpURLConnection does the right thing in this case when receiving a "Connection: close" header and update here. Give me a day or so.

**Comment by sirthias** [comment]
The question of how we deal with with early responses does not depend on the implementation of Java's HttpURLConnection, but it does of course provide an interesting data point.
Looking forward to your findings.

**Comment by RichardBradley** [comment]
Yes -- adding "Connection: close" to the response when the server detects that the request is only part-read (or not read at all) fixes this issue entirely :-)

(See [this commit](https://github.com/RichardBradley/akka-connection-reuse/commit/eb378a1d1d4e092ed21d7c996f034c968d365739) -- the tests now pass)

I'm happy that #16898 will fix this issue.

Thanks very much

**Comment by sirthias** [comment]
Great. Thanks a lot again for reporting and digging into this!
Your contributions are very much appreciated!

**Comment by RichardBradley** [comment]
Thanks for your help.

(I'm glad the fix here is so simple and effective. I had managed to convince myself this was some huge morass of interlocking bugs and incompatibilities.)
